overlord/snapstate/storehelpers.go

```diff
@@ -623,14 +623,46 @@ func installCandidates(st *state.State, names []string, channel string, user *au
 		return nil, err
 	}
 
+	enforcedSets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return nil, err
+	}
+
 	actions := make([]*store.SnapAction, len(names))
 	for i, name := range names {
-		actions[i] = &store.SnapAction{
+		action := &store.SnapAction{
 			Action:       "install",
 			InstanceName: name,
 			// the desired channel
 			Channel: channel,
 		}
+
+		if enforcedSets != nil {
+			// check for invalid presence first to have a list of sets where it's invalid
+			invalidForValSets, err := enforcedSets.CheckPresenceInvalid(naming.Snap(name))
+			if err != nil {
+				if _, ok := err.(*snapasserts.PresenceConstraintError); !ok {
+					return nil, err
+				} // else presence is optional or required, carry on
+			}
+			if len(invalidForValSets) > 0 {
+				return nil, fmt.Errorf("cannot install snap %q due to enforcing rules of validation set %s", name, strings.Join(invalidForValSets, ","))
+			}
+			requiredValSets, requiredRevision, err := enforcedSets.CheckPresenceRequired(naming.Snap(name))
+			if err != nil {
+				return nil, err
+			}
+			if len(requiredValSets) > 0 {
+				setActionValidationSets(action, requiredValSets)
+				if !requiredRevision.Unset() {
+					action.Revision = requiredRevision
+					// channel cannot be present if revision is set (store would
+					// respond with revision-conflict error).
+					action.Channel = ""
```

Comment:
wondering if we need to do this given that store will do this anyway?

Reply:
No strong opinion, but FWTW not having this here makes the tests a little bit confusing since channel: "stable" shows up in the resulting fakeBackend.ops because we default to "stable" in InstallMany.

Reply:
shouldn't we do it in the helper though now if Revision gets set?  that's the behavior we get from store all the time too, so it's never wrong

Reply:
Indeed, makes sense to do it in the helper, thanks.